plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/stream_info.py

```diff
@@ -1373,111 +1396,138 @@ def load_stream_info(self, video_id):
+
+                    if (age_gate_enabled
+                            and playability.get('desktopLegacyAgeGateReason')):
+                        abort = True
                         break
-                    if any(why in compare_reason for why in abort_reasons):
+                    elif status == 'LIVE_STREAM_OFFLINE':
                         abort = True
                         break
+                    elif status == 'OK':
+                        break
+                    elif status in {
+                        'AGE_CHECK_REQUIRED',
+                        'AGE_VERIFICATION_REQUIRED',
+                        'CONTENT_CHECK_REQUIRED',
+                        'LOGIN_REQUIRED',
+                        'CONTENT_NOT_AVAILABLE_IN_THIS_APP',
+                        'ERROR',
+                        'UNPLAYABLE',
+                    }:
+                        log_warning(
+                            'Failed to retrieve video info'
+                            '\n\tStatus:   {status}'
+                            '\n\tReason:   {reason}'
+                            '\n\tvideo_id: |{video_id}|'
+                            '\n\tClient:   |{client}|'
+                            '\n\tAuth:     |{auth}|'
+                            .format(
+                                status=status,
+                                reason=reason or 'UNKNOWN',
+                                video_id=video_id,
+                                client=_client['_name'],
+                                auth=_client.get('_has_auth', False),
+                            )
+                        )
+                        compare_reason = reason.lower()
+                        if any(why in compare_reason for why in reauth_reasons):
+                            if has_access_token:
+                                client_data['_auth_required'] = True
+                                restart = True
+                            break
+                        if any(why in compare_reason for why in retry_reasons):
+                            continue
+                        if any(why in compare_reason for why in skip_reasons):
+                            break
+                        if any(why in compare_reason for why in abort_reasons):
+                            abort = True
+                            break
+                    else:
+                        log_debug(
+                            'Unknown playabilityStatus in player response'
+                            '\n\tplayabilityStatus: {0}'
+                            .format(playability)
+                        )
                 else:
-                    self._context.log_debug(
-                        'Unknown playabilityStatus in player response:\n|{0}|'
-                        .format(playability)
-                    )
+                    break
+                if not restart:
+                    break
+                restart = False
```

Comment:
Should you not wait a bit here?

Reply:
Yes, maybe, but not exactly, I think :)

Yes - it's probably good practice.

Maybe - it's a service that can easily handle the requests, and other 3rd party clients and even the official clients perform similar requests without waiting.

But not exactly - a wait time, if used, would probably be better placed in the existing `for` loop that makes the requests, rather than in the new `while` loop, that was added to reset the existing `for` loop.

The master branch that is pushed to this repository has already been updated, and it is a bit of a pain to reset it back to make a non-conflicting commit. Can do so if necessary, but otherwise can instead add this in the next version?

Reply:
Actually on second thought I don't think there should be any wait time between requests.

I can fill the connection table of my local router with parallel requests (numbering in the thousands) without seeing any issues with remote responses. There are defined error responses for too many requests but i can't seem to hit that limit so the 2-4 requests made here shouldn't be a problem.

Given that the official clients do something similar but using QUIC, I think the logic is that it is less resource intensive to respond to multiple requests using the same connection rather than waiting and potentially opening new connections.

Reply:
I don't really care about the service not being able to handle, more that if the loop keeps going, it might cause Kodi to freeze?

Reply:
Well if the loop keeps going it will probably make Kodi unresponsive regardless of any time delay that was added.

I don't believe the loop can keep going because of the different way in which the service responds to unauthenticated vs authenticated requests.

Just in case the request responses do change in the future, I will add something like the following to the next release.

```
Subject: [PATCH] Avoid potential player request loop in case future request responses do not distinguish between (un)authenticated requests
--- a/resources/lib/youtube_plugin/youtube/helper/stream_info.py
+++ b/resources/lib/youtube_plugin/youtube/helper/stream_info.py
@@ -1414,7 +1414,7 @@ class StreamInfo(YouTubeRequestClient):
             if name == 'ask' and use_mpd and not ask_for_quality:
                 continue
 
-            restart = False
+            restart = None
             while 1:
                 for client_name in clients:
                     _client = self.build_client(client_name, client_data)
@@ -1480,7 +1480,10 @@ class StreamInfo(YouTubeRequestClient):
                         )
                         compare_reason = _reason.lower()
                         if any(why in compare_reason for why in reauth_reasons):
-                            if has_access_token:
+                            if client_data.get('_auth_required'):
+                                restart = False
+                                abort = True
+                            elif restart is None and has_access_token:
                                 client_data['_auth_required'] = True
                                 restart = True
                             break
```